### PR TITLE
bpf: tests: pktgen infra for tunneling + GENEVE-DSR test

### PR DESCRIPTION
**Pull request by julianwiedmann**
Add the necessary bits so that pktgen can build tunnel encapsulation headers (UDP, VXLAN, GENEVE). Exercise it with some initial tests - more to come.

### CODE REVIEW

File: bpf/tests/pktgen.h [hunk 1]
```diff
@@ -834,7 +886,20 @@ void pktgen__finish(const struct pktgen *builder)
...
+			 */
+			if (layer_off >= MAX_PACKET_OFF - sizeof(struct udphdr))
+				return;
+
+			udp_layer = ctx_data(builder->ctx) + layer_off;
+			if ((void *)udp_layer + sizeof(struct udphdr) >
+			    ctx_data_end(builder->ctx))
+				return;
+
+			if (i + 1 >= PKT_BUILDER_LAYERS)
```
Comment (gentoo-root) [inline bpf/tests/pktgen.h]: Is it needed? Won't it return because the loop finishes?

File: bpf/tests/pktgen.h [hunk 2]
```diff
@@ -914,6 +1042,36 @@ void pktgen__finish(const struct pktgen *builder)
...
+			geneve_layer = ctx_data(builder->ctx) + layer_off;
+			if ((void *)geneve_layer + sizeof(struct genevehdr) >
+			    ctx_data_end(builder->ctx))
+				return;
+
+			if (i + 1 >= PKT_BUILDER_LAYERS)
+				return;
+			break;
+
+			switch (builder->layers[i + 1]) {
```
Comment (gentoo-root) [inline bpf/tests/pktgen.h:1121]: Looks like this code is unreachable because of the break above.

**Review by nbusseneau — APPROVED** [review]

File: bpf/tests/pktgen.h (hunk 1 above)
Comment (julianwiedmann) [reply bpf/tests/pktgen.h]: :+1: copy & pasted from other cases that require the check before accessing the next layer, but for UDP it's not needed.

File: bpf/tests/pktgen.h (hunk 2 above)
Comment (julianwiedmann) [reply bpf/tests/pktgen.h:1121]: ha, thanks for catching that. I wonder if that explains some of the failures I was seeing in my test backlog ...

**Review by gentoo-root — APPROVED** [review]

**Pull request merged**